src/logics/Logic.cc

```diff
@@ -1356,8 +1355,10 @@ bool Logic::varsubstitute(PTRef root, const Map<PTRef, PtAsgn, PTRefHash> & subs
 #ifdef SIMPLIFY_DEBUG
                     printf("  %s -> %s\n", printTerm(t[i]), printTerm(gen_sub[t[i]]));
 #endif
+                    if (hasSub and appearsInUF(t[i])) {
+                        decreaseUFAppearanceCount(t[i]);
+                    }
```

Comment:
> Is it correct to always decrement here? What if such term is being substituted in a non-UF context? Or that cannot happen?

I think it can happen.  In my quick tests I could verify this.  I'll add a check there that `isUF(tr)`, I hope that solves the issue.

> And if the count is decremented for the child here, shouldn't it be incremented for the term that is being substituted in its place?

This should already happen when the new term is constructed with the standard means, if I get what you mean.  I'll maybe have to come up with an example to verify this.

Reply:
> This should already happen when the new term is constructed with the standard means, if I get what you mean. 

Yes, you are right. That's fine then.

> I think it can happen. In my quick tests I could verify this. I'll add a check there that isUF(tr), I hope that solves the issue.

Maybe this is not necessary, since the substitution should actually remove the term from everywhere, so maybe this is a not a problem. But I probably still don't understand what the original issue with these substitutions was, so don't mind me :)

Reply:
> the substitution should actually remove the term from everywhere

I think that's not true, the substitution is not on the term that appears in UF, but on something else that results in a uf-appearance disappearing.

Reply:
Now I agree with what you say above, and disagree what I wrote as a reply.  Maybe this was not the original issue, but instead the problem was somehow related to whether a term has ever appeared in UF or if it has appeared at some point but right now does not appear.  I think I'll try it that way, it might make the code a little more compact.